flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ModifyOperation.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.operations;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.api.QueryConfig;
+import org.apache.flink.table.api.TableSchema;
+
+import java.util.List;
+
+/**
+ * A {@link TableOperation} that describes the DML queries such as e.g. INSERT or conversion to a
+ * DataStream.
+ *
+ * <p>A tree of {@link QueryOperation} with a {@link ModifyOperation} on top
+ * represents a runnable query that can be transformed into a graph of
+ * {@link org.apache.flink.streaming.api.transformations.StreamTransformation}
+ * via {@link org.apache.flink.table.planner.Planner#translate(List, QueryConfig)}
+ *
+ * <p>It extends from {@link QueryOperation} to enable applying subsequent
+ * {@link QueryOperation}s on top. In other words it makes it possible to have an output
+ * transformation as a middle step of a relational query. This is not supported yet.
+ *
+ * @see QueryOperation
+ */
+@Internal
+public abstract class ModifyOperation extends QueryOperation {
```

Comment:
I think we should not let this class extends from `QueryOperation`, it makes the class hierarchy hard to understand. There are plenty of ways to achieve goal of applying subsequent `QueryOperation` on it.

Reply:
Also, how about change the name to  `SinkOperation` ?

Reply:
We wanted to be more inline with SQL terms. You have DQ(uery)L, DM(odify)L, DC(ontrol)L, and DD(efinition)L.

Reply:
As the javadoc states, the `Modify` operation extends from `Query` operation because modify operations can also produce output and just be part of a SQL query. 

We saw a lot of requests to allow having transformations applied to `Sink`s in a `DataStream` API. This class structure prepares for such use cases.

Right now this is a forbidden operation, because the `getTableSchema` method throws `UnsupportedOperation`, so you just cannot apply subsequent `Operations` on top of this one.

Reply:
Keep align with SQL sounds good. 

Regarding to the extends question, I agree we can support reading from a sinked table. But conceptually, this should be separated as two things, 1) write data to a sink. 2) read data from the underlying storage of the table. 
We can provide some syntax sugar for this kind of requirements, but not with the hierarchy. If this class is align with SQL keyword, does "DML is a DQL" sounds weird to you?

Reply:
I didn't mean `write data to a sink and then read it again`. What I meant is something as follows:
you write data to sink and the output of the operation is e.g. the number of rows stored (this is the standard SQL behavior), or e.g. id of record in an external DB.

Ad. "does DML is a DQL" actually there are quite a few different opinions. The most prominent is that SELECT(DQL) is also a DML, but if you think about a query like `SELECT * FROM INSERT INTO` it may be considered also the other way round.

Reply:
I had a quick chat with @twalthr about that. Even though I think this is a valid thing to have at some point in time we don't need to have it for now. 

As this is an internal representation we can change that later.

Reply:
ok, sounds good to me